src/health/health_event_loop.c

```diff
@@ -348,391 +547,755 @@ static void health_event_loop_for_host(RRDHOST *host, bool apply_hibernation_del
+    ARAL *ar;
+} health_config_s = { 0 };
+
+static struct health_cmd health_deq_cmd(void)
+{
+    struct health_cmd ret = { 0 };
+    struct health_cmd *to_free = NULL;
+
+    spinlock_lock(&health_config_s.cmd_queue_lock);
+    if(health_config_s.cmd_base) {
+        struct health_cmd *t = health_config_s.cmd_base;
+        DOUBLE_LINKED_LIST_REMOVE_ITEM_UNSAFE(health_config_s.cmd_base, t, prev, next);
+        ret = *t;
+        to_free = t;
+    }
+    else {
+        ret.opcode = HEALTH_NOOP;
+    }
+    spinlock_unlock(&health_config_s.cmd_queue_lock);
+    aral_freez(health_config_s.ar, to_free);
 
-        time_t now = now_realtime_sec();
-        bool apply_hibernation_delay = false;
-        time_t next_run = now + health_globals.config.run_at_least_every_seconds;
+    return ret;
+}
 
-        if (unlikely(check_if_resumed_from_suspension())) {
-            apply_hibernation_delay = true;
+static void health_enq_cmd(struct health_cmd *cmd)
+{
+    struct health_cmd *t = aral_mallocz(health_config_s.ar);
+    *t = *cmd;
+    t->prev = t->next = NULL;
 
-            nd_log(NDLS_DAEMON, NDLP_NOTICE,
-                   "Postponing alarm checks for %"PRId32" seconds, "
-                   "because it seems that the system was just resumed from suspension.",
-                   (int32_t)health_globals.config.postpone_alarms_during_hibernation_for_seconds);
-            schedule_node_state_update(localhost, 0);
-        }
+    spinlock_lock(&health_config_s.cmd_queue_lock);
+    DOUBLE_LINKED_LIST_APPEND_ITEM_UNSAFE(health_config_s.cmd_base, t, prev, next);
+    spinlock_unlock(&health_config_s.cmd_queue_lock);
 
-        if (unlikely(silencers->all_alarms && silencers->stype == STYPE_DISABLE_ALARMS)) {
-            static int logged=0;
-            if (!logged) {
-                nd_log(NDLS_DAEMON, NDLP_DEBUG,
-                       "Skipping health checks, because all alarms are disabled via API command.");
-                logged = 1;
-            }
-        }
+    (void) uv_async_send(&health_config_s.async);
+}
 
-        worker_is_busy(WORKER_HEALTH_JOB_RRD_LOCK);
-        uint64_t loop = __atomic_add_fetch(&health_evloop_iteration, 1, __ATOMIC_RELAXED);
+struct job_list_t {
+    health_job_type_t job_type;
+    int pending;
+    int running;
+    int max_threads;
+    Pvoid_t JudyL;
+    Word_t count;
+};
 
-        RRDHOST *host;
-        dfe_start_reentrant(rrdhost_root_index, host) {
-            if(unlikely(!service_running(SERVICE_HEALTH)))
-                break;
+static void async_cb(uv_async_t *handle)
+{
+    uv_stop(handle->loop);
+    uv_update_time(handle->loop);
+}
 
-            health_event_loop_for_host(host, apply_hibernation_delay, now, &next_run);
-        }
-        dfe_done(host);
+#define TIMER_PERIOD_MS (5000)
 
-        if(unlikely(!service_running(SERVICE_HEALTH)))
-            break;
+static void timer_cb(uv_timer_t *handle)
+{
+    uv_stop(handle->loop);
+    uv_update_time(handle->loop);
```

Comment:
I am letting this comment here for we do not forget. If next commits do not use the commented code, please, remove them.